src/normalize/normalize.rs

```diff
@@ -0,0 +1,421 @@
+
+
+/// A type representing a build ID note.
+///
+/// In the ELF file, this header is typically followed by the variable sized
+/// build ID.
+#[repr(C)]
+struct BuildIdNote {
+    /// ELF note header.
+    header: Elf64_Nhdr,
+    /// NUL terminated string representing the name.
+    name: [u8; 4],
+}
+
+// SAFETY: `BuildIdNote` is valid for any bit pattern.
+unsafe impl crate::util::Pod for BuildIdNote {}
+
+
+/// Attempt to read an ELF binary's build ID.
+// TODO: Currently look up is always performed based on section name, but there
+//       is also the possibility of iterating notes and checking checking
+//       Elf64_Nhdr.n_type for NT_GNU_BUILD_ID, specifically.
+fn read_build_id(path: &Path) -> Result<Option<Vec<u8>>> {
+    let build_id_section = ".note.gnu.build-id";
+    let file = File::open(path)?;
+    let parser = ElfParser::open_file(file)?;
+
+    // The build ID is contained in the `.note.gnu.build-id` section. See
+    // elf(5).
+    if let Ok(idx) = parser.find_section(build_id_section) {
+        // SANITY: We just found the index so the section should always be
+        //         found.
+        let shdr = parser.section_headers()?.get(idx).unwrap();
+        if shdr.sh_type != elf::types::SHT_NOTE {
+            warn!(
+                "build ID section {build_id_section} of {} is of unsupported type ({})",
+                path.display(),
+                shdr.sh_type
+            );
+            return Ok(None)
+        }
+
+        // SANITY: We just found the index so the section should always be
+        //         found.
+        let mut bytes = parser.section_data(idx).unwrap();
+        let header = bytes.read_pod_ref::<BuildIdNote>().ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidData,
+                "failed to read build ID section header",
+            )
+        })?;
+        if &header.name != b"GNU\0" {
+            warn!(
+                "encountered unsupported build ID type {:?}; ignoring",
+                header.name
+            );
+            Ok(None)
+        } else {
+            // Every byte following the header is part of the build ID.
+            let build_id = bytes.to_vec();
+            Ok(Some(build_id))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+
+/// Normalize a virtual address belonging to an ELF file represented by the
+/// provided [`MapsEntry`].
+fn normalize_elf_addr(virt_addr: Addr, entry: &MapsEntry) -> Result<Addr> {
+    let file_off = virt_addr - entry.range.start + entry.offset as usize;
+    let parser = ElfParser::open(&entry.path)?;
+    let phdrs = parser.program_headers()?;
+    let addr = phdrs
+        .iter()
+        .find_map(|phdr| {
+            if phdr.p_type == elf::types::PT_LOAD {
+                if (phdr.p_offset..phdr.p_offset + phdr.p_memsz).contains(&(file_off as u64)) {
+                    return Some(file_off + phdr.p_vaddr as usize - phdr.p_offset as usize)
```

Comment:
nit: to avoid (somewhat hypothetical)  overflow, it's best to subtract p_offset before adding p_vaddr

Reply:
Sure, done.